### PR TITLE
Check for subpath error in container status instead of pod events

### PR DESCRIPTION
**Pull request by msau42**
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
Events can be unreliable. Check for errors in container status instead.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially addresses #71570

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/storage/testsuites/subpath.go [hunk 1]
```diff
@@ -587,16 +586,35 @@ func testPodFailSubpathError(f *framework.Framework, pod *v1.Pod, errorMsg strin
...
+				} else if status.State.Waiting != nil {
+					if status.State.Waiting.Reason == "CreateContainerConfigError" &&
+						strings.Contains(status.State.Waiting.Message, "subPath") {
+						return true, nil
+					}
+					return false, nil
+				}
+			}
+		}
+		return false, nil
```
Comment (davidz627) [inline test/e2e/storage/testsuites/subpath.go]: This case we never found a status corresponding to the container name. Just want to make sure simply retrying is the right behavior here

File: test/e2e/storage/testsuites/subpath.go [hunk 2]
```diff
@@ -587,16 +586,35 @@ func testPodFailSubpathError(f *framework.Framework, pod *v1.Pod, errorMsg strin
...
+					return false, fmt.Errorf("subpath container unexpectedly became running")
+				} else if status.State.Terminated != nil {
+					return false, fmt.Errorf("subpath container unexpectedly terminated")
+				} else if status.State.Waiting != nil {
+					if status.State.Waiting.Reason == "CreateContainerConfigError" &&
+						strings.Contains(status.State.Waiting.Message, "subPath") {
+						return true, nil
+					}
+					return false, nil
+				}
```
[the remark this answers is not in the document]
Comment (msau42) [reply test/e2e/storage/testsuites/subpath.go]: I haven't confirmed, but I think it's possible that container status is not initially populated/empty.

File: test/e2e/storage/testsuites/subpath.go (hunk 1 above)
Comment (msau42) [reply test/e2e/storage/testsuites/subpath.go]: This is similar to the case where container status may not be initially populated

File: test/e2e/storage/testsuites/subpath.go [hunk 3]
```diff
@@ -587,16 +586,35 @@ func testPodFailSubpathError(f *framework.Framework, pod *v1.Pod, errorMsg strin
...
-	Expect(err).NotTo(HaveOccurred(), "while waiting for failed event to occur")
+func waitForPodSubpathError(f *framework.Framework, pod *v1.Pod) error {
+	return wait.PollImmediate(framework.Poll, framework.PodStartTimeout, func() (bool, error) {
+		pod, err := f.ClientSet.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, status := range pod.Status.ContainerStatuses {
+			// 0 is the container that uses subpath
+			if status.Name == pod.Spec.Containers[0].Name {
```
[the remark this answers is not in the document]
Comment (msau42) [reply test/e2e/storage/testsuites/subpath.go]: done

File: test/e2e/storage/testsuites/subpath.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/storage/testsuites/subpath.go]: nit: use case statement for clarity, if you'd like

**Pull request merged**